beluga_ros/include/beluga_ros/particle_cloud.hpp

```diff
@@ -91,6 +91,32 @@ struct almost_equal_to<Sophus::SE2<Scalar>> {
   const Scalar angular_resolution;  ///< Resolution for rotational coordinates, in radians.
 };
 
+/// std::equal_to equivalent specialized for SE(3) types, with user resolutions.
+template <typename Scalar>
+struct almost_equal_to<Sophus::SE3<Scalar>> {
+  /// Constructs near equality functor.
+  /**
+   * \param _linear_resolution Resolution for translational coordinates, in meters.
+   * \param _angular_resolution Resolution for rotational coordinates, in radians.
+   */
+  explicit almost_equal_to(Scalar _linear_resolution, Scalar _angular_resolution)
+      : linear_resolution(_linear_resolution), angular_resolution(_angular_resolution) {}
+
+  /// Compares `a` and `b` for near equality.
+  bool operator()(const Sophus::SE3<Scalar>& a, const Sophus::SE3<Scalar>& b) const {
+    using std::abs, std::atan, std::tan;
+    const Eigen::Vector3<Scalar> angles = a.so3().log() - b.so3().log();
+    return (abs(a.translation().x() - b.translation().x()) < linear_resolution) &&
+           (abs(a.translation().y() - b.translation().y()) < linear_resolution) &&
+           (abs(a.translation().z() - b.translation().z()) < linear_resolution) &&
+           (abs(atan(tan(angles.x()))) < angular_resolution) && (abs(atan(tan(angles.y()))) < angular_resolution) &&
+           (abs(atan(tan(angles.z()))) < angular_resolution);
```

Comment:
@pvela2017 I wonder if it wouldn't be simpler (and more accurate) to just:

```suggestion
    const Sophus::SE3<Scalar> diff  = b.inverse() * a;
    const Eigen::Vector3<Scalar> angles = diff.so3().log();
    return (abs(diff.translation().x()) < linear_resolution) &&
           (abs(diff.translation().y()) < linear_resolution) &&
           (abs(diff.translation().z()) < linear_resolution) &&
           (abs(atan(tan(angles.x()))) < angular_resolution) && (abs(atan(tan(angles.y()))) < angular_resolution) &&
           (abs(atan(tan(angles.z()))) < angular_resolution);
```

Reply:
Indeed @glpuga. A difference in SO3 logspace will only approximate the logmap of an SO3 difference within a small neighborhood. Which means we botched the SO2 comparison as well: we are `atan(tan(...))` normalizing the difference, not the individual terms :eyes:

Reply:
Would it be something like this?

```cpp
template <typename Scalar>
struct almost_equal_to<Sophus::SE2<Scalar>> {
  /// Constructs near equality functor.
  /**
   * \param _linear_resolution Resolution for translational coordinates, in meters.
   * \param _angular_resolution Resolution for rotational coordinates, in radians.
   */
  explicit almost_equal_to(Scalar _linear_resolution, Scalar _angular_resolution)
      : linear_resolution(_linear_resolution), angular_resolution(_angular_resolution) {}

  /// Compares `a` and `b` for near equality.
  bool operator()(const Sophus::SE2<Scalar>& a, const Sophus::SE2<Scalar>& b) const {
    using std::abs;
    const Scalar theta = (a * b.inverse()).log().x();
    return (abs(a.translation().x() - b.translation().x()) < linear_resolution) &&
           (abs(a.translation().y() - b.translation().y()) < linear_resolution) &&
           (abs(theta) < angular_resolution);
  }
```